packages/hadron-build/commands/upload.js

```diff
@@ -1,139 +1,466 @@
+      return downloadCenter;
+    });
+
+  cli.info('Uploading assets to download center…');
+
+  await checkAssetsExist(
+    assetsToUpload.map((asset) => {
+      return asset.path;
+    })
   );
 
+  const uploads = assetsToUpload.map(async(asset) => {
+    cli.info(
+      `${asset.name}: upload to download center started (path: ${path.relative(
+        root,
+        asset.path
+      )}).`
+    );
+    if (!dryRun) {
+      await uploadAsset(channel, asset);
+    }
+    cli.info(`${asset.name}: upload to download center completed.`);
+  });
+
   await Promise.all(uploads);
 }
 
-exports.command = 'upload [options]';
+/**
+ *
+ * @param {'stable' | 'beta'} channel
+ */
+async function getLatestRelease(channel = 'stable') {
+  const octokit = new Octokit({
+    auth: process.env.GITHUB_TOKEN
+  });
+
+  const page = 1;
+
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    let releases = [];
+
+    try {
+      const { data } = await octokit.request(
+        'GET /repos/{owner}/{repo}/releases',
+        {
+          owner: 'mongodb-js',
+          repo: 'compass',
+          per_page: 100,
+          page
+        }
+      );
+      releases = data;
+    } catch (err) {
+      cli.warn(`Failed to fetch releases: ${err.message}`);
+    }
+
+    // We ran out of releases or failed to fetch
+    if (releases.length === 0) {
+      return null;
+    }
+
+    const latestRelease = releases.find((release) => {
+      return (
+        !release.draft &&
+        (channel === 'beta'
+          ? isBeta(release.tag_name)
+          : isStable(release.tag_name))
+      );
+    });
+
+    if (latestRelease) {
+      return latestRelease;
+    }
+
+    page++;
+  }
+}
+
+function getDiffFromAssertionError(message) {
+  return message.split(/\+ actual.+?\n/)[1];
+}
+
+async function getLatestReleaseVersions(channel = 'stable') {
+  const release = await getLatestRelease(channel);
+  if (!release) {
+    throw new Error(`Couldn't find latest release for ${channel} channel`);
+  }
+  const manifest = release.assets.find((asset) => {
+    return asset.name === 'manifest.json';
+  });
+  if (!manifest) {
+    throw new Error(`No manifest found in the release for ${channel} channel`);
+  }
+  const content = JSON.parse(
+    (await download(manifest.browser_download_url)).toString()
+  );
+  return content;
+}
+
+async function updateManifest(dryRun) {
+  cli.info('Downloading current manifest');
+  const currentManifest = await downloadManifest();
+
+  const versions = (
+    await Promise.all(
+      ['stable', 'beta'].map((channel) => {
+        cli.info(`Looking for the latest release for the ${channel} channel`);
+        return getLatestReleaseVersions(channel);
+      })
+    )
+  ).flat();
+
+  const newManifest = {
+    ...currentManifest,
+    versions
+  };
+
+  try {
+    deepStrictEqual(currentManifest, newManifest);
+    cli.warn('Skipping upload: manifests are identical');
+    return;
```

Comment:
Not sure if we should fail here instead or not

Reply:
I like that, seems a legit case for example if we publish/unpublish old versions

Reply:
Though this code confused me a tons, cause it throws in the "happy path", would it be ok to extract the check to a function that doesn't throw?

```
isDeepStrictEqual =  (obj1, obj2) => {
   try { deepStrictEqual(obj1, obj2); return true; }  catch { return false; }
}

if (isDeepStrictEqual(m1, m2)) {
   return;
}

// actually upload

```